src/NUnitFramework/framework/Attributes/TestCaseSourceAttribute.cs

```diff
@@ -179,67 +184,87 @@ public IEnumerable<ITestCaseData> GetTestCasesFor(MethodInfo method)
         {
             Type sourceType = this.SourceType;
             if (sourceType == null)
                 sourceType = method.ReflectedType;
 
+            // Handle Type implementing IEnumerable separately
             if (SourceName == null)
-                return Reflect.Construct(sourceType, _sourceConstructorParameters) as IEnumerable;
+                return Reflect.Construct(sourceType) as IEnumerable;
 
             MemberInfo[] members = sourceType.GetMember(SourceName,
                     BindingFlags.Public | BindingFlags.NonPublic | BindingFlags.Static | BindingFlags.Instance);
+
             if (members.Length == 1)
             {
                 MemberInfo member = members[0];
-                object sourceobject = Reflect.Construct(sourceType, _sourceConstructorParameters);
 
                 var field = member as FieldInfo;
                 if (field != null)
-                    return (IEnumerable)field.GetValue(sourceobject);
+                    return field.IsStatic
+                        ? (IEnumerable)field.GetValue(null)
+                        : SourceMustBeStaticError();
 
                 var property = member as PropertyInfo;
                 if (property != null)
-                    return (IEnumerable)property.GetValue(sourceobject, null);
+                    return property.GetGetMethod(true).IsStatic
+                        ? (IEnumerable)property.GetValue(null, null)
+                        : SourceMustBeStaticError();
 
                 var m = member as MethodInfo;
                 if (m != null)
-                    return (IEnumerable)m.Invoke(sourceobject, null);
+                    return m.IsStatic
+                        ? (IEnumerable)m.Invoke(null, null)
+                        : SourceMustBeStaticError();
             }
 
             return null;
         }
-        #endregion
 
-        #region ITestBuilder Members
-
-        /// <summary>
-        /// Construct one or more TestMethods from a given MethodInfo,
-        /// using available parameter data.
-        /// </summary>
-        /// <param name="method">The MethodInfo for which tests are to be constructed.</param>
-        /// <param name="suite">The suite to which the tests will be added.</param>
-        /// <returns>One or more TestMethods</returns>
-        public IEnumerable<TestMethod> BuildFrom(MethodInfo method, Test suite)
+        private static IEnumerable SourceMustBeStaticError()
         {
-            List<TestMethod> tests = new List<TestMethod>();
-
-            foreach (ParameterSet parms in GetTestCasesFor(method))
-                tests.Add(_builder.BuildTestMethod(method, suite, parms));
-
-            return tests;
+            var parms = new ParameterSet();
+            parms.RunState = RunState.NotRunnable;
+            parms.Properties.Set(PropertyNames.SkipReason, "The name specified on a TestCaseSourceAttribute must refer to a static field, property or method.");
```

Comment:
In both constructors, the parameter in question is `sourceName`. Should we be more specific in this message and say something like "The sourceName parameter on the TestCaseSourceAttribute..."?